src/core/MIGRATION.md

```diff
@@ -604,9 +604,138 @@ It is generally a much greater challenge preparing legacy browser-side code for
 
 To complicate matters further, a significant amount of the business logic in Kibana's client-side code exists inside the `ui/public` directory (aka ui modules), and all of that must be migrated as well. Unlike the server-side code where the order in which you migrated plugins was not particularly important, it's important that UI modules be addressed as soon as possible.
 
-Also unlike the server-side migration, we won't concern ourselves with creating shimmed plugin definitions that then get copied over to complete the migration.
+Because usage of angular and `ui/public` modules varies widely between legacy plugins, there is no "one size fits all" solution to migrating your browser-side code to the new platform. The best place to start is by checking with the platform team to help identify the best migration path for your particular plugin.
 
-### Move UI modules into plugins
+That said, we've seen a series of patterns emerge as teams begin migrating browser code. In practice, most migrations will follow a path that looks something like this:
+
+#### 1. Create a plugin definition file
+
+We've found that doing this right away helps you start thinking about your plugin in terms of lifecycle methods and services, which makes the rest of the migration process feel more natural. It also forces you to identify which actions "kick off" your plugin, since you'll need to execute those when the `setup/start` methods are called.
+
+This definition isn't going to do much for us just yet, but as we get further into the process, we will gradually start returning contracts from our `setup` and `start` methods, while also injecting dependencies as arguments to these methods.
+
+```ts
+// public/plugin.ts
+import { CoreSetup, CoreStart, Plugin } from '../../../../core/public';
+import { FooSetup, FooStart } from '../../../../legacy/core_plugins/foo/public';
+
+/** @public */
+export type DemoSetup = {}
+/** @public */
+export type DemoStart = {}
+
+/** @internal */
+export interface DemoSetupPlugins {
+  foo: FooSetup;
+}
+/** @internal */
+export interface DemoStartPlugins {
+  foo: FooStart;
+}
+
+/** @internal */
+export class DemoPublicPlugin implements Plugin<DemoSetup, DemoStart, DemoSetupPlugins, DemoStartPlugins> {
+  public setup(core: CoreSetup, plugins: DemoSetupPlugins): DemoSetup {
```

Comment:
* `Demo*Plugins` are plugins that we have dependencies on
* `DemoSetupPlugins` are the contracts returned by each plugin's `setup` phase
* `DemoStartPlugins` are the contracts returned by each plugin's `start` phase

Right?

I think that this should be a comment in this code sample - it will be useful for when people copy that into their own plugins.

Reply:
Correct, but in looking at this again I will update those names to be aligned with the ones in CONVENTIONS.md: https://github.com/elastic/kibana/blob/70a979a59d8ab61f2115788d29e47d0530e2dea9/src/core/CONVENTIONS.md#L92-L100

In reality the names of the *deps interfaces shouldn't matter too much since they should be internal to your plugin. What's more important is the `FooSetup` and `FooStart` naming, which will be exported. But either way I will update for consistency.